src/passes/pass.cpp

```diff
@@ -647,6 +647,11 @@ void PassRunner::runOnFunction(Function* func) {
 }
 
 void PassRunner::doAdd(std::unique_ptr<Pass> pass) {
+  if (!pass->supports(wasm->features) && !options.forceUnsupportedPasses) {
+    std::cerr << "[PassRunner] Skipping '" << pass->name
+              << "' because module has unsupported features" << std::endl;
+    return;
```

Comment:
I think we should have a `Fatal() << "..` here, as silently disabling passes may end up breaking things - e.g. if someone does `--flatten --somethingthatneedsflatten` then flatten would be skipped and the next pass either break or not do all it should. I appreciate that you've disabled the flatten-depending passes too, but we might forget that with new passes or other users might forget.

I also think without a fatal error here that things like asyncify and wasm2js may break, unless we already have an error there for exceptions (without that, it would skip i64 lowering, for example, and likely break later). Overall it seems safest to just error here.

Reply:
That sounds fine, but then `wasm-opt -O4 -all` will break every time, whether the module actually uses exception or not, because `-O4` includes flatten and `-all` enables everything. Is this gonna be OK?

Reply:
I agree that skipping unsupported passes sounds bad, but having `wasm-opt -O4 -all` break also sounds bad. Maybe `-all` should only enable "complete" features (for some definition of complete) and we should have another flag like `-experimental` to actually enable all features?

Reply:
Not sure about that either. Flatten and dependent passes, except I64toI32Lowering that's required for wasm2js, only run when given `-O4` (unless you enable them independently), and making EH work on them is not super high priority for now. Should we make a feature experimental every time there is even a single pass that's not supported?

Reply:
Yeah I guess one reasonable cutoff would be to say that all optimization (i.e. semantics-preserving) passes need to be supported. But maybe all optimization passes that can be run at -O3 is enough. Not strictly relevant, but I'm surprised Flatten is not run at O3, too.

Reply:
Many good points here!

Can we try to run a pass that is incompatible with exceptions, and only error if we actually see a `Try`? That way `-O4 -all` would work on most code.